GH-2107: Use a progress bar to show the overall upload status

kinow opened the pull request:
GitHub issue resolved #2107

Pull request Description:

Display the statuses of file uploads near the top / visible area, so users do not need to scroll the page to track the progress of their uploads.

----

 - [x] Tests are included.
 - [ ] Documentation change and updates are provided for the [Apache Jena website](https://github.com/apache/jena-site/)
 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx, or if in JIRA, JENA-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

----

See the [Apache Jena "Contributing" guide](https://github.com/apache/jena/blob/main/CONTRIBUTING.md).

kinow commented:
Hi @neumarcx ,

What do you think of this first prototype?

![image](https://github.com/apache/jena/assets/304786/ca1f7655-3920-4d53-8f5f-1ca370fce004)

![image](https://github.com/apache/jena/assets/304786/9bba7012-593c-4602-a2a4-c54b763da8d4)

(I used the JS console to both confirm there are no errors, and to hide some uploads so we can see what it would look like -- also developing for mobile/smaller viewport by default.)

Not fully functional as I have to hook another reactive field to compute the number of successes and failures. Any other metric we should include?

Once we have agreed on the UI prototype and how it should behave, I will make it work correctly, add a test, and a changelog entry.

Thanks!

neumarcx commented:
very nice Bruno, yes I think that's all we need. Not sure that we need them to be multi functional value boxes though.

how about the following mockup. "Success" and / or "Failure"  may actually not need to be shown prior to the upload.

just an additional idea.

![Jena](https://github.com/apache/jena/assets/10403482/9e1cd8ba-4da9-487a-a0cf-3c8cbcf069ee)

neumarcx commented:
> Hi @neumarcx !
> 
> +1 for "Upload status"!
> 
> > how about the following mockup. "Success" and / or "Failure" may actually not need to be shown prior to the upload.
> 
> My first prototype didn't show the status by default, then I realized it would "move things in the UI". That's considered bad for UX [1](#user-content-fn-1-0a2d933431e1bda521c96cf9c0e0b774), but more importantly can impact performance negatively [2](#user-content-fn-2-0a2d933431e1bda521c96cf9c0e0b774), so it'd be nice if we could find a permanent place that doesn't move things too much.
> 
> And I didn't use colors because the red would attract your eyes (at least it does to me when I open a page like that), even if it says 0 (imagine Twitter with the notification symbol enabled by default showing 0, or GitHub's notification icon). So instead of users slowly reading the page, their first eye scanning would be to look at the red, and then proceed to use the page[3](#user-content-fn-3-0a2d933431e1bda521c96cf9c0e0b774). Green would have the same effect, and it has also color-blindness issues [4](#user-content-fn-4-0a2d933431e1bda521c96cf9c0e0b774).
> 
> I tried to place each number in a fixed location using table columns so users would know where to find them, once they learn it. But a single line without a table could work too.
> 
> What do you think? Maybe something like your prototype but with the text only? Any thoughts @afs, @rvesse ?
> 
> ## Footnotes
> 1. that's CLS, and I hate it because some sites (Twitter, DeviantArt, ...) and apps (Zoom, Discord) have this, that when you have your mouse over a button and something happens and the DOM elements shift, and you end up clicking on something else: https://javascript.plainenglish.io/3-ways-to-fix-cumulative-layout-shift-53df8bea1e0f?gi=e99977032785 [↩](#user-content-fnref-1-0a2d933431e1bda521c96cf9c0e0b774)
> 2. because it moves other HTML DOM elements, which might cause JS to have to re-render code like Vue.js components, and redraw things [↩](#user-content-fnref-2-0a2d933431e1bda521c96cf9c0e0b774)
> 3. related to visual weight & visual direction: https://www.smashingmagazine.com/2014/12/design-principles-visual-weight-direction/#visual-weight [↩](#user-content-fnref-3-0a2d933431e1bda521c96cf9c0e0b774)
> 4. https://www.smashingmagazine.com/2016/06/improving-color-accessibility-for-color-blind-users/#6-color-combinations [↩](#user-content-fnref-4-0a2d933431e1bda521c96cf9c0e0b774)

Thank you for the comprehensive motivation for you choices and I can follow your position on my idea with the references.

I also see you point with regards to the Cumulative Layout Shift. I don't have an immediate opinion on it but can intuitively sense the issue. The UI might even feel broken as the user might misread the design.

Could you please elaborate the one liner idea? May main concern with the multi functional  table columns was that they indicate more functionality than is never used. 
I may have miss interpreted how the status table works. Will the table expand with multiple errors or always it always be juts one line? One table header row and one table value row?

kinow commented:
Sorry, I haven't had time to look at your previous comment, @neumarcx . I'd say we continue discussing to get a good prototype, then I hook up the JS code to make the prototype into actual working code, add unit tests, and then mark as ready for review.

neumarcx commented:
Ok good @kinow keep me in the loop. I guess I have to close the issue at some point.

Where do I check out your current version of the work? The current jena github repo is already tagged for Jena 5. I presume that's not the version it will be available for but it will be available for version Jena 4.10.

kinow commented:
I think this might go into Jena 5, unless we finish it soon and want to backport it to jena4 (`main` is jena5 now, https://github.com/apache/jena/blob/cb23ddcd974df1c110332326a116ba88bc90d475/pom.xml#L29, see jena4 branch for Jena4 related work). The changes are in the branch of this repo, better-upload-ux.

I will keep updating this PR as I have time, and you should get some GitHub notifications and/or maybe emails. Once we merge this PR it should close the issue automatically.

kinow commented:
Hi @neumarcx ,

Coming back at this one :wave:

I was thinking in moving the status information "inside" the upload component, instead of being a section above the upload form.

And instead of a table with three columns, after thinking a bit I thought maybe we could use progress bar like the rest of the uploads, where you have an upload bar per file.

Here's the first preview. I am prototyping it, but pretend there are files selected and the upload is happening :slightly_smiling_face: 

![image](https://github.com/apache/jena/assets/304786/dcfa856c-e225-4239-aaba-358cda5dc215)

The success bar can be green or blue, but I thought green could be confusing because of the select all button (although blue is also mixing with the upload all button). Initially the bars were empty, then I added 30% and 20%, and then I added the 3/10 and 2/10 to have the same information we had before (to answer a question like "how many files were selected).

WDYT?

kinow commented:
Refactored to match what we discussed on Twitter, @neumarcx. Edited commit, and did manual tests. It's now missing just the unit/e2e tests, and then it'll be ready to go. Some screenshots of what it looks like in action.

Blank page when you open the page to add data:

![image](https://github.com/apache/jena/assets/304786/b22d418f-5384-4efb-a684-9c6612fb21f0)

Selecting one invalid file, and two valid files.

![image](https://github.com/apache/jena/assets/304786/82e400d2-aa8e-480d-9bf0-9aa63a6bd8af)

Hitting upload all:

![image](https://github.com/apache/jena/assets/304786/56364cbf-7286-4769-99ed-668da87667a0)

If you remove one of the files, it's reactive, so it adjusts the progress bar accordingly:

![image](https://github.com/apache/jena/assets/304786/26f4df5f-c48e-4caf-b381-9dc8fe0c6da5)

And if you add a new file, it again updates correctly.

![image](https://github.com/apache/jena/assets/304786/5c3b80cc-a9d5-4710-b44e-44e01cd9ebdd)

I think that will indeed help users who need to upload multiple files at once. Next step the tests (when I find time for that :+1:)

kinow commented:
Tests added, @neumarcx, tested `yarn test:unit`, `yarn lint`, and `yarn run test:e2e` locally and all passed.

The test is an e2e test, that loads three files. There is a Cypress “interceptor” that, as the name says, intercepts network requests to the URL to handle the upload, and the test-interceptor fails every odd-time it is called.

There are tests for no files selected, showing an empty progress bar (so no content shift, or very minimum, which is good!), and a test that adds three files, and confirms we get 2/3 success, and 1/3 failure, and that the progress bars (4 in total, an overall progress, and one for each file selected) are displayed correctly.

:+1:

afs commented:
Good great - tested using the uploads page in a few ways.

kinow commented:
> Good great - tested using the uploads page in a few ways.

Thanks Andy!

afs commented:
Presumably, the branch "better-upload-ux" can be deleted?

kinow commented:
Yes, forgot to do that, sorry. Branch deleted.
